File: src/dungeonbreak_narrative/escape_the_dungeon/narrative/fame.py
"""Deterministic Fame context formula."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Mapping


def _clamp(value: float, low: float, high: float) -> float:
    return max(low, min(high, value))


@dataclass(frozen=True)
class FameFormulaInput:
    current_fame: float
    effort_spent: float
    room_vector: Mapping[str, float]
    action_novelty: float
    risk_level: float
    momentum: float
    has_broadcast_skill: bool


@dataclass(frozen=True)
class FameFormulaResult:
    gain: float
    base_gain: float
    context_multiplier: float
    diminishing_factor: float
    components: dict[str, float]


def compute_fame_gain(data: FameFormulaInput) -> FameFormulaResult:
    projection = float(data.room_vector.get("Projection", 0.0))
    levity = float(data.room_vector.get("Levity", 0.0))
    direction = float(data.room_vector.get("Direction", 0.0))
    survival = float(data.room_vector.get("Survival", 0.0))

    room_interest = _clamp((0.6 * projection) + (0.4 * levity) + (0.25 * direction) + (0.2 * survival), -0.6, 1.6)
    novelty = _clamp(float(data.action_novelty), 0.0, 1.6)
    risk = _clamp(float(data.risk_level), 0.0, 1.4)
    momentum_bonus = _clamp(float(data.momentum) * 0.04, 0.0, 0.30)
    skill_bonus = 0.15 if data.has_broadcast_skill else 0.0

    context_multiplier = max(
        0.15,
        1.0
        + (0.45 * room_interest)
        + (0.30 * novelty)
        + (0.25 * risk)
        + momentum_bonus
        + skill_bonus,
    )
    effort_factor = max(0.0, float(data.effort_spent) / 10.0)
    base_gain = 1.0 * effort_factor
    diminishing = 1.0 / (1.0 + (max(0.0, float(data.current_fame)) / 120.0))
    gain = round(max(0.0, base_gain * context_multiplier * diminishing), 4)
    return FameFormulaResult(
        gain=gain,
        base_gain=base_gain,
        context_multiplier=context_multiplier,
        diminishing_factor=diminishing,
        components={
            "room_interest": room_interest,
            "novelty": novelty,
            "risk": risk,
            "momentum_bonus": momentum_bonus,
            "skill_bonus": skill_bonus,
        },
    )

File: docs-site/lib/escape-the-dungeon/narrative/fame.ts
import { clamp } from "@/lib/escape-the-dungeon/core/types";

export interface FameFormulaInput {
  currentFame: number;
  effortSpent: number;
  roomVector: Record<string, number>;
  actionNovelty: number;
  riskLevel: number;
  momentum: number;
  hasBroadcastSkill: boolean;
}

export interface FameFormulaResult {
  gain: number;
  baseGain: number;
  contextMultiplier: number;
  diminishingFactor: number;
  components: Record<string, number>;
}

export const computeFameGain = (input: FameFormulaInput): FameFormulaResult => {
  const projection = Number(input.roomVector.Projection ?? 0);
  const levity = Number(input.roomVector.Levity ?? 0);
  const direction = Number(input.roomVector.Direction ?? 0);
  const survival = Number(input.roomVector.Survival ?? 0);

  const roomInterest = clamp(
    0.6 * projection + 0.4 * levity + 0.25 * direction + 0.2 * survival,
    -0.6,
    1.6,
  );
  const novelty = clamp(Number(input.actionNovelty), 0, 1.6);
  const risk = clamp(Number(input.riskLevel), 0, 1.4);
  const momentumBonus = clamp(Number(input.momentum) * 0.04, 0, 0.3);
  const skillBonus = input.hasBroadcastSkill ? 0.15 : 0;

  const contextMultiplier = Math.max(
    0.15,
    1 + 0.45 * roomInterest + 0.3 * novelty + 0.25 * risk + momentumBonus + skillBonus,
  );

  const effortFactor = Math.max(0, Number(input.effortSpent) / 10);
  const baseGain = 1 * effortFactor;
  const diminishingFactor = 1 / (1 + Math.max(0, Number(input.currentFame)) / 120);
  const gain = Number(Math.max(0, baseGain * contextMultiplier * diminishingFactor).toFixed(4));

  return {
    gain,
    baseGain,
    contextMultiplier,
    diminishingFactor,
    components: {
      roomInterest,
      novelty,
      risk,
      momentumBonus,
      skillBonus,
    },
  };
};
